demos/howtos/myula.py

```diff
@@ -0,0 +1,287 @@
+#
+# :math:`g_{\texttt{smoothing_strength}}` is continuously differentiable with :math:`1/\texttt{smoothing_strength}`-Lipschitz gradient and s.t
+#
+# .. math::
+#       \nabla g_{\texttt{smoothing_strength}} (x) = (x- \operatorname{prox}_g^{\texttt{smoothing_strength}} (x))/\texttt{smoothing_strength}
+#
+# with
+#
+# .. math::
+#       \operatorname{prox}_g^{\texttt{smoothing_strength}} (x) = \operatorname{argmin}_z \|x-z \|_2^2/2\texttt{smoothing_strength} + g(z)
+#
+# See https://link.springer.com/chapter/10.1007/978-3-319-48311-5_31 for more details.
+#
+# MYULA consists in applying ULA to a smoothed target distribution. It reads
+#
+# .. math::
+#       \begin{align*}
+#       x_{k+1} &= x_k + \texttt{scale} \nabla \log \pi_{\texttt{smoothing_strength}}(x_k |y) + \sqrt{2 \texttt{scale}} z_{k+1}\\
+#       &= x_k + \texttt{scale} \nabla \log \pi(y | x_k) + \texttt{scale} \nabla \log \pi_{\texttt{smoothing_strength}}(x_k) + \sqrt{2 \texttt{scale}} z_{k+1}\\
+#       &= x_k + \texttt{scale} \nabla \log \pi(y | x_k) - \texttt{scale} (x_k - \operatorname{prox}_g^{\texttt{smoothing_strength}} (x_k))/{\texttt{smoothing_strength}} + \sqrt{2 \texttt{scale}} z_{k+1}.
+#       \end{align*}
+#
+# where :math:`\texttt{smoothing_strength}` corresponds to the smoothing strength of :math:`g`.
+#
+# To illustrate MYULA, we will consider :math:`g(x) = \texttt{strength_reg} \  TV(x) = \texttt{strength_reg} \|\nabla x \|_{2, 1}`,
+# where :math:`\texttt{strength_reg}` is the regularization parameter which
+# controls the regularization strength induced by TV.
+
+# %%
+# Bayesian model definition
+# -------------------------
+# Then consider the following Bayesian Inverse Problem:
+#
+# .. math::
+#    \begin{align*}
+#    \mathbf{x} &\sim \exp (- \texttt{strength_reg} \|\nabla x \|_{2,1})\\
+#    \mathbf{y}_{obs} &\sim \mathcal{N}(\mathbf{A}\mathbf{x}, \texttt{sigma2}\,\mathbf{I}) \ ,
+#    \end{align*}
+#
+# with :math:`\texttt{sigma2}=0.05^2`.
+
+# %%
+# Likelihood definition
+# ---------------------
+# We first specify the data distribution as follows:
+sigma2=0.05**2
+y=cuqi.distribution.Gaussian(A, sigma2)
+# %%
+# Then we can define the likelihood with
+likelihood=y(y=y_obs)
+
+# %%
+# RestorationPrior and MoreauYoshidaPrior
+# ---------------------------------------
+# To apply MYULA, we need to define the Moreau-Yoshida prior 
+# :math:`\pi_{\texttt{smoothing_stength}}(x)`.
+# Evaluating this surrogate prior is doable but too intensive from
+# a computational point of view as it requires to solve an optimization problem.
+# However to apply MYULA, we only require access to
+# :math:`\operatorname{prox}_{\texttt{strength_reg}\ TV}^{\texttt{smoothing_strength}}`.
+# :math:`\operatorname{prox}_{\texttt{strength_reg}\ TV}^{\texttt{smoothing_strength}}`
+# is a denoising operator, which takes a signal as input and returns a less noisy
+# signal. In CUQIPy, we talk about restoration operators. It is an operator at the
```

Comment:
I don't think we use "restoration operator" anywhere else, and think we should rephrase here to avoid calling it an operator.

Reply:
Inside RestorationPrior the term "restorator" is used. I suspect that is equivalent to what is called "restoration operator" here?  I suggest consistently using "restorator" to avoid the term "operator" which can have multiple interpretations.

Reply:
Thanks Jakob! That is a good point. We clarified it in the documentation of myula.py stating that restoration operators are restorators and denoisers are denoising operators which are a special case of restorators.